### DISCUSSION TITLE
How fontforge modifies the value of the boundingBox？

### DISCUSSION THREAD

**Discussion by zhenzi0322** [fontforge/fontforge#5144] [General]
How fontforge modifies the value of the boundingBox？

I would like to know how the boundingBox is calculated？

![97](https://user-images.githubusercontent.com/34839719/202409618-5347acef-9166-4fd0-a688-05f359b66534.svg)

**Comment by skef** [comment]
Of the whole font or an individual glyph? And how are you accessing the data, through the UI or the native scripting language or python?

**Reply by zhenzi0322** [reply]
I am trying to use python script to generate fonts, but the preview of the generated font file is not good. 

As shown below.

![image](https://user-images.githubusercontent.com/34839719/202597912-079e6878-84d6-4387-aea7-87cd9041aead.png)

**Reply by skef** [reply]
Assuming you haven't added kerning, the spacing between characters is determined by these two amounts:

1. The space between the origin (x=0) and the left side of the glyph
2. The horizontal advance. 

From python the horizontal advance is the value in `glyph.width`, which can be read or edited. It's normally set to an integer (and off-hand I can't remember whether it can be set to anything else).

However, it may be easiest to start by playing with the `font.autoWidth()` method, or just trying that out in the FontForge UI. There's some more (partially out of date but still useful) information about all this here: https://fontforge.org/docs/tutorial/editexample5.html and here: https://fontforge.org/docs/ui/dialogs/autowidth.html

**Reply by skef** [reply]
Oh, for individual glyphs in the FontForge UI you can just drag the horizontal advance line around. It's the one line (usually) to the right of the glyph paths.

**Reply by zhenzi0322** [reply]
```python
import fontforge


def GlyphSeparation(onLeft, onRight, context):
    # the goal is to give a weighted average that expresses the visual
    #  separation between two glyphs when they are placed so their bounding
    #  boxes are adjacent. The separation between two rectangles would be 0
    #  While the separation between "T" and "o" would be fairly large
    # The trick is to guess a good weighting function. My guess is that
    #  things that look close are more important than those which look far
    #  So "T" and "O" should be dominated by the crossbar of the "T"...
    #
    # Find the area the two glyphs have in common (so when comparing "o" and
    #  "l" or "g", ignore the ascenders and descenders since that's outside
    #  the range of "o" and won't affect its visual proximity.
    imin_y = max(onRight.iminY, onLeft.iminY);
    imax_y = min(onRight.imaxY, onLeft.imaxY);
    #
    # Some glyph combinations will have no overlap -- the grave accent and the
    # letter "a" will have none. So they don't interact visually.
    if imax_y < imin_y:
        return (0)
    #
    # Otherwise compute some sort of weighted average of the separations between
    # the two glyphs at various heights.
    tot, cnt = 0, 0
    j = imin_y
    while j <= imax_y:
        if onRight.left[j] < 32767 and onLeft.right[j] > -32767:
            # beware of gaps such as those in "i" or "aaccute"
            # a gap has a left or right value which is huge
            # so ignore any such, again it doesn't contribute to the
            # visual separation.
            sep = onRight.left[j] - onLeft.right[j]
            weight = 1.0 / (sep + context.denom)
            weight *= weight
            tot += weight * sep
            cnt += weight
        j += 1
    if cnt != 0:
        tot /= cnt
    return (int(round(tot)));


fontforge.registerGlyphSeparationHook(GlyphSeparation)

IMPORT_OPTIONS = ('removeoverlap', 'correctdir')
font = fontforge.font()


def add_glyph(unicode, svg_path):
    g = font.createMappedChar(unicode)
    g.importOutlines(svg_path, IMPORT_OPTIONS)
    g.removeOverlap()


def main():
    add_glyph(unicode=67, svg_path="static/svg/67.svg")
    add_glyph(unicode=97, svg_path="static/svg/97.svg")
    add_glyph(unicode=104, svg_path="static/svg/104.svg")
    add_glyph(unicode=105, svg_path="static/svg/105.svg")
    add_glyph(unicode=110, svg_path="static/svg/110.svg")

    font.generate("test.ttf")
    print("success")


if __name__ == '__main__':
    main()
```

Generate the following:
![image](https://user-images.githubusercontent.com/34839719/202601551-caa848ea-a1c1-4c22-8f1c-9c60a2e88336.png)

The glyph spacing was not well aligned before. How do I need to adjust it to make the font alignment neat?

**Reply by skef** [reply]
First off you may be happier with `g.importOutlines(svg_path, scale=False, correctdir=True)` and then running `g.removeOverlap()`. That `IMPORT_OPTIONs` tuple won't do what you expect. 

Looks like we don't currently use the SVG Viewbox to set the horizontal advance width, unfortunately. @ctrlcctrlv do we have standard advice for this case?

**Reply by ctrlcctrlv** [reply]
Hmm. I guess you could create a deprecated SVG 1.1 font?

**Reply by zhenzi0322** [reply]
```python
import fontforge


def GlyphSeparation(onLeft, onRight, context):
    imin_y = max(onRight.iminY, onLeft.iminY)
    imax_y = min(onRight.imaxY, onLeft.imaxY)
    if imax_y < imin_y:
        return (0)
    tot, cnt = 0, 0
    j = imin_y
    while j <= imax_y:
        if onRight.left[j] < 32767 and onLeft.right[j] > -32767:
            sep = onRight.left[j] - onLeft.right[j]
            weight = 1.0 / (sep + context.denom)
            weight *= weight
            tot += weight * sep
            cnt += weight
        j += 1
    if cnt != 0:
        tot /= cnt
    return (int(round(tot)))


fontforge.registerGlyphSeparationHook(GlyphSeparation)

font = fontforge.font()
font.ascent = 336
font.descent = 80
font.em = 416


def add_glyph(unicode, svg_path):
    g = font.createMappedChar(unicode)
    g.importOutlines(svg_path, scale=False, correctdir=True)
    g.removeOverlap()


def main():
    add_glyph(unicode=67, svg_path="static/svg/67.svg")
    add_glyph(unicode=97, svg_path="static/svg/97.svg")
    add_glyph(unicode=104, svg_path="static/svg/104.svg")
    add_glyph(unicode=105, svg_path="static/svg/105.svg")
    add_glyph(unicode=110, svg_path="static/svg/110.svg")

    font.generate("test.ttf")
    print("success")


if __name__ == '__main__':
    main()
```
![image](https://user-images.githubusercontent.com/34839719/202605379-58508e27-611b-496a-9d1c-dd768ea18d7b.png)

[test3.zip](https://github.com/fontforge/fontforge/files/10037274/test3.zip)

**Reply by zhenzi0322** [reply]
> First off you may be happier with `g.importOutlines(svg_path, scale=False, correctdir=True)` and then running `g.removeOverlap()`. That `IMPORT_OPTIONs` tuple won't do what you expect.
> 
> Looks like we don't currently use the SVG Viewbox to set the horizontal advance width, unfortunately. @ctrlcctrlv do we have standard advice for this case?

Still hasn't achieved the effect of word spacing

**Reply by ctrlcctrlv** [reply]
Wait, _word_ spacing?

**Reply by zhenzi0322** [reply]
> Wait, _word_ spacing?

Right, the glyphs are not aligned

**Reply by zhenzi0322** [reply]
I used a 256x256 bitmap to turn out the svg, and then did the generated ttf

**Reply by zhenzi0322** [reply]
I want to create an effect like the following:

![image](https://user-images.githubusercontent.com/34839719/202606299-ec31148a-885e-4a75-8857-e1a9cabd1e42.png)

**Reply by ctrlcctrlv** [reply]
Did FontForge make these SVG?

**Reply by zhenzi0322** [reply]
> Did FontForge make these SVG?

No, I exported the svg via potrace

**Reply by ctrlcctrlv** [reply]
That won't work how you want. Your SVG files should have the same metrics as your font in FontForge.

**Reply by zhenzi0322** [reply]
If it has to be the same metrics, is there no way to generate the font file via python script?

**Reply by zhenzi0322** [reply]
There should be a way to adjust it, I just want to adjust the spacing of the glyphs, the spacing of the glyphs can not be modified by py script generation? Or is it possible to adjust the spacing by modifying the svg

**Reply by ctrlcctrlv** [reply]
Did you generate with `--tight` from potrace? Don't do that.

**Reply by zhenzi0322** [reply]
Yes. And how to do it well?

**Reply by ctrlcctrlv** [reply]
Remove that flag for starters :-)

**Reply by zhenzi0322** [reply]
I use the following command：
```
potrace xxx.jpg -b svg -o xxx.svg --tight --flat
```

**Reply by ctrlcctrlv** [reply]
Remove `--tight`.

**Reply by zhenzi0322** [reply]
移除之后生成svg文件内容类似如下：
```svg
<svg height="256.000000pt" preserveAspectRatio="xMidYMid meet" version="1.0" viewBox="0 0 256.000000 256.000000" width="256.000000pt" xmlns="http://www.w3.org/2000/svg"><g fill="#000000" stroke="none" transform="translate(0.000000,256.000000) scale(0.100000,-0.100000)"><path d="M1198 1944 c-83 -20 -148 -56 -216 -119 -159 -147 -235 -364 -248 -705 -17 -421 75 -681 291 -825 147 -97 344 -110 504 -33 76 36 177 141 217 224 39 82 69 208 79 337 l7 87 -114 0 -115 0 -7 -97 c-12 -181 -45 -274 -124 -345 -152 -138 -377 -51 -451 174 -38 116 -46 194 -46 423 1 303 29 447 111 570 69 104 176 151 291 127 130 -27 204 -144 220 -348 l6 -84 110 0 110 0 -6 98 c-17 273 -157 465 -379 517 -77 18 -164 18 -240 -1z"/></g></svg>
```

The preview of the generated ttf effect is as follows.

![image](https://user-images.githubusercontent.com/34839719/202610456-3157f6b3-7cac-4729-8221-825da4ffe032.png)

**Reply by zhenzi0322** [reply]
```python
import fontforge

font = fontforge.font()
font.ascent = 336
font.descent = 80
font.em = 416

def add_glyph(unicode, svg_path):
    print(unicode)
    g = font.createMappedChar(unicode)
    g.importOutlines(svg_path, scale=False, correctdir=True)
    g.removeOverlap()


def main():
    add_glyph(unicode=67, svg_path="static/svg/67.svg")
    add_glyph(unicode=97, svg_path="static/svg/97.svg")
    add_glyph(unicode=104, svg_path="static/svg/104.svg")
    add_glyph(unicode=105, svg_path="static/svg/105.svg")
    add_glyph(unicode=110, svg_path="static/svg/110.svg")
    add_glyph(unicode=111, svg_path="static/svg/111.svg")
    add_glyph(unicode=116, svg_path="static/svg/116.svg")

    font.generate("test.ttf")
    print("success")

if __name__ == '__main__':
    main()
```

**Reply by zhenzi0322** [reply]
Here is the svg file for remove tight.

[svg.zip](https://github.com/fontforge/fontforge/files/10037537/svg.zip)

**Reply by ctrlcctrlv** [reply]
OK that's much better and closer to what you want. Now you just have the problem @skef mentioned, not one problem confounding the other problem. I don't know how to do it in a script (check the API) but in the UI there is a box called "infer glyph width".

One other thing to note: your SVG files demand `xMidYMid meet` for `preserveAspectRatio`. I doubt we support this, which is causing the stretching, I think.

![image](https://user-images.githubusercontent.com/838783/202687068-f9139baa-111b-4a8b-87f7-138ae599d8b1.png)

**Reply by ctrlcctrlv** [reply]
Oh and by the way @zhenzi0322. I actually suggest that you consider a different approach to this entirely, and use FontForge's internal integration with Potrace rather than making SVG's outside of it, which just adds another layer of abstraction.

**Reply by ctrlcctrlv** [reply]
(哦！還有一個想法給@zhenzi0322：完全採取不同的方法；使用Potrace與FontForge的內部集成，而不是使用FontForge之外Potrace輸出SVG文件（這只是增加了另一層抽象）｡)

**Reply by zhenzi0322** [reply]
Ok, how does the internal integration work?

**Reply by ctrlcctrlv** [reply]
https://fontforge.org/docs/scripting/python/fontforge.html#fontforge.glyph.autoTrace

**Reply by zhenzi0322** [reply]
Does fontforge support importing bitmaps and then generating font files

**Reply by zhenzi0322** [reply]
> https://fontforge.org/docs/scripting/python/fontforge.html#fontforge.glyph.autoTrace

How does this work? Do you have sample code?

**Reply by ctrlcctrlv** [reply]
Yes...you should really read the documentation and not ask every question you have until you try and get stuck.

**Reply by zhenzi0322** [reply]
Ok

**Comment by skef** [comment]
The FontForge import mechanism uses the SVG baseline as the glyph baseline. It looks like your SVG input files might not be positioned and/or scaled consistently. If they aren't you'll either have to fix them by hand after importing or make them consistent prior to importing.

**Reply by zhenzi0322** [reply]
> The FontForge import mechanism uses the SVG baseline as the glyph baseline. It looks like your SVG input files might not be positioned and/or scaled consistently. If they aren't you'll either have to fix them by hand after importing or make them consistent prior to importing.

They are currently consistent and both are exported via 256x256 bitmaps

**Comment by zhenzi0322** [comment]
Here are my svg files and script files

https://github.com/fontforge/fontforge/files/10037274/test3.zip
